src/oidcop/oidc/token.py

```diff
@@ -352,6 +363,281 @@ def post_parse_request(
+                and request["requested_token_type"] in config["policy"]
+            ):
+                callable = config["policy"][request["requested_token_type"]]["callable"]
+                kwargs = config["policy"][request["requested_token_type"]]["kwargs"]
+            else:
+                callable = config["policy"][""]["callable"]
+                kwargs = config["policy"][""]["kwargs"]
+
+            fn = importer(callable)
+            return fn(request, context=_context, subject_token=token, **kwargs)
+
+        except Exception:
+            return self.error_cls(
+                error="server_error",
+                error_description="Internal server error"
+            )
+
+    def token_exchange_response(self, token):
+        response_args = {}
+        response_args["access_token"] = token.value
+        response_args["scope"] = token.scope
+        response_args["issued_token_type"] = token.token_class
+        response_args["expires_in"] = token.usage_rules["expires_in"]
+        response_args["token_type"] = "bearer"
+
+        return TokenExchangeResponse(**response_args)
+
+    def process_request(self, request, **kwargs):
+        _context = self.endpoint.server_get("endpoint_context")
+        _mngr = _context.session_manager
+        try:
+            _session_info = _mngr.get_session_info_by_token(
+                request["subject_token"], grant=True
+            )
+        except ToOld:
+            logger.error("Subject token has expired.")
+            return self.error_cls(
+                error="invalid_request",
+                error_description="Subject token has expired"
+            )
+        except (KeyError, UnknownToken):
+            logger.error("Subject token invalid.")
+            return self.error_cls(
+                error="invalid_request",
+                error_description="Subject token invalid"
+            )
+
+        token = _mngr.find_token(_session_info["session_id"], request["subject_token"])
+
+        grant = _session_info["grant"]
+
+        _requested_token_type = request.get("requested_token_type",
+                                           "urn:ietf:params:oauth:token-type:access_token")
+
+        _token_class = _requested_token_type.split(":")[-1]
+        if _token_class == "access_token":
+            _token_type = _token_class
+        else:
+            _token_type = None
+
+        sid = _session_info["session_id"]
+        if request["client_id"] != _session_info["client_id"]:
+            authn_event = _session_info["grant"].authentication_event
+            _token_usage_rules = _context.authz.usage_rules(request["client_id"])
+            _exp_in = _token_usage_rules["refresh_token"].get("expires_in")
+            if _exp_in and "valid_until" in authn_event:
+                authn_event["valid_until"] = utc_time_sans_frac() + _exp_in
+
+            sid = _mngr.create_session(
+                authn_event=authn_event,
+                auth_req=request,
```

Comment:
this isn't an authentication request, it's a token request, perhaps this should be `None`?